[PT Run] System plugin: Add IP and MAC

htcfreek opened the pull request:
## Summary of the Pull Request

**What is this about:**
This adds two new commands to show the IPs and MAC addresses of each network connection interface. To search for them on global queries the search term has to start with `ip`, `mac` or `address`.

![image](https://user-images.githubusercontent.com/61519853/158835341-6c855660-0743-41c2-81b5-0adcd1902d40.png)

**Note:** ~The new image has a not so good quality at the moment. I have asked @niels9001 for a better one. But I think the image should not block this PR from merging.~ The symbol is updated. (But not the screenshot.)

**How does someone test / validate:** 
Test with local build.

## Quality Checklist

- [x] **Linked issue:** #16968
- [x] **Communication:** I've discussed this with core contributors in the issue. 
- [x] **Tests:** Added/updated and all pass
- [x] **Installer:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [x] **Dev docs:** Added/ updated
- [x] **User docs:** see #16630
- [x] **Binaries:** Any new files are added to WXS / YML
   - [ ] No new binaries
   - [ ] [YML for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [x] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.

htcfreek commented:
@niels9001
Can help me with creating better images for the new results? The have a bad quality.

htcfreek commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/SystemCommandResultContext.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Microsoft.PowerToys.Run.Plugin.System.Components
+{
+    internal class SystemCommandResultContext
```
```suggestion
    internal class SystemCommandResult
```

htcfreek commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/SystemCommandResultContext.cs:
```diff
@@ -0,0 +1,26 @@
+        public string Data { get; set; }
+    }
+
+    internal enum SystemCommandResultType
```
```suggestion
    internal enum CommandType
```

niels9001 commented:
> @niels9001 Can help me with creating better images for the new results? The have a bad quality.

Doesn't it make sense to just use the Settings plugin icon as this is part of that plugin? My worry is that it might become challenging to find relevant icons for all extensions at some point. (e.g. a network card icon for IP address might be confusing to users)

htcfreek commented:
It's system plugin angd not settings plugin. And we don't have a useful generic icon in this plugin.

htcfreek commented:
> > @niels9001 Can help me with creating better images for the new results? The have a bad quality.
> 
> Doesn't it make sense to just use the Settings plugin icon as this is part of that plugin? My worry is that it might become challenging to find relevant icons for all extensions at some point. (e.g. a network card icon for IP address might be confusing to users)  

It's system plugin angd not settings plugin. And we don't have a useful generic icon in this plugin. 

For me the settings icon don't feels right. But we can use the "network adapter" icon for both. That would reduce the number of different icons.

I plan to add some network commands like `ipconfig` too. Which icon would you use for that?

htcfreek commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/SystemPluginContext.cs:
```diff
@@ -0,0 +1,26 @@
+    internal enum ResultContextType
+    {
+        Command,
+        IpResult,
+        MacResult,
```
NetworkAdapterInfo

htcfreek commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/ResultHelper.cs:
```diff
@@ -0,0 +1,91 @@
+
+            if (contextData.Type == ResultContextType.IpResult)
+            {
+                contextMenu.Add(new ContextMenuResult()
```
remove

jsoref commented:
@htcfreek:
`etwork` is indeed from `\network`, just add it here:
https://github.com/microsoft/PowerToys/blob/main/.github/actions/spell-check/patterns.txt#L48

I'm sorry. At some point, I'm going to probably move the magic that handles `\n`/`\r`/`\t` into https://github.com/check-spelling/spell-check-this/blob/main/.github/actions/spelling/patterns.txt and out of the core defaults. I think it's right about 95% of the time in general, and maybe 50-75% of the time when dealing w/ Windows systems, but that isn't really good enough.

Not sure about the other bit, can you clarify?

htcfreek commented:
@niels9001
Regarding the icon some things have changed. We now use the same icon for mac and ip. I only need a good icon version for the `network adapter` symbol.

jsoref commented:
@htcfreek: So, the [base dictionary](https://raw.githubusercontent.com/check-spelling/check-spelling/dictionary-20200211/dict.txt) appears to have:

```
BPS
bps
BPSS
KBPS
kbps
MBPS
mbps
```

(I have no idea what `BPSS` is.)

The dictionary's `mbps` entry is covering your `Mbps`. You can verify by applying these changes:

```diff
diff --git a/.github/actions/spell-check/expect.txt b/.github/actions/spell-check/expect.txt
index 0279f0ca3..81714abde 100644
--- a/.github/actions/spell-check/expect.txt
+++ b/.github/actions/spell-check/expect.txt
@@ -1185,3 +1185,2 @@ maxversiontested
 Mbits
-Mbps
 MBs
@@ -2391,2 +2390,2 @@ Zoneszonabletester
 Zonev
-zzz
\ No newline at end of file
+zzz
diff --git a/.github/actions/spell-check/patterns.txt b/.github/actions/spell-check/patterns.txt
index 66274dccd..a9bc2dca1 100644
--- a/.github/actions/spell-check/patterns.txt
+++ b/.github/actions/spell-check/patterns.txt
@@ -51,2 +51,3 @@ TestCase\("[^"]+"
 \\netstandard
+\\network
 \\notifications
diff --git a/.github/actions/spell-check/reject.txt b/.github/actions/spell-check/reject.txt
new file mode 100644
index 000000000..9e4575656
--- /dev/null
+++ b/.github/actions/spell-check/reject.txt
@@ -0,0 +1 @@
+mbps
```

You want the _patterns_ change and _expect_ changes, but obviously for this PR you wouldn't want to add the _reject_ file, it's just for demonstration purposes.

The notable thing about the dictionary is that it doesn't have `gbps`/`GBPS`, the dictionary is "old", so that was clearly too far into its future for it to be a concept that was included.

htcfreek commented:
> @htcfreek: So, the [base dictionary](https://raw.githubusercontent.com/check-spelling/check-spelling/dictionary-20200211/dict.txt) appears to have:
> 
> ```
> BPS
> bps
> BPSS
> KBPS
> kbps
> MBPS
> mbps
> ```
> 
> (I have no idea what `BPSS` is.)
> 
> The dictionary's `mbps` entry is covering your `Mbps`. You can verify by applying these changes:
> 
> ```diff
> diff --git a/.github/actions/spell-check/expect.txt b/.github/actions/spell-check/expect.txt
> index 0279f0ca3..81714abde 100644
> --- a/.github/actions/spell-check/expect.txt
> +++ b/.github/actions/spell-check/expect.txt
> @@ -1185,3 +1185,2 @@ maxversiontested
>  Mbits
> -Mbps
>  MBs
> @@ -2391,2 +2390,2 @@ Zoneszonabletester
>  Zonev
> -zzz
> \ No newline at end of file
> +zzz
> diff --git a/.github/actions/spell-check/patterns.txt b/.github/actions/spell-check/patterns.txt
> index 66274dccd..a9bc2dca1 100644
> --- a/.github/actions/spell-check/patterns.txt
> +++ b/.github/actions/spell-check/patterns.txt
> @@ -51,2 +51,3 @@ TestCase\("[^"]+"
>  \\netstandard
> +\\network
>  \\notifications
> diff --git a/.github/actions/spell-check/reject.txt b/.github/actions/spell-check/reject.txt
> new file mode 100644
> index 000000000..9e4575656
> --- /dev/null
> +++ b/.github/actions/spell-check/reject.txt
> @@ -0,0 +1 @@
> +mbps
> ```
> 
> You want the _patterns_ change and _expect_ changes, but obviously for this PR you wouldn't want to add the _reject_ file, it's just for demonstration purposes.
> 
> The notable thing about the dictionary is that it doesn't have `gbps`/`GBPS`, the dictionary is "old", so that was clearly too far into its future for it to be a concept that was included.

A had to add `Gbps` to `expect.txt`. Is this missing in the base dictionary?

jsoref commented:
Yeah, I know it's confusing. Dictionaries can evolve, but in general, any dictionary will be a snapshot in time. And in order to not surprise people, I haven't evolved the dictionary check-spelling uses.

I do think I'll probably at some point make at least one revision to the dictionary. Including `GBPS`/`gbps` and maybe even `TBPS`/`tbps` seems like something I might want to do -- I'm not entirely certain I'd want to do that, as `tbps` could easily be a typo for `tbsp`. Making the call about what to add into a dictionary is actually hard. It's a lot easier

What would happen is that there'd be a newer version of the dictionary available, and people could opt-in to use it. And then much further into the future, I might change the default, so at an upgrade, people could get that newer dictionary, but opt to use the older one.

---

I think my general advice is to not try to predict what the spell checker will suggest putting into `expect.txt`. If you're concerned that something _should_ be a word regardless of what happens, you could add a line for it into a `.txt` file in: https://github.com/microsoft/PowerToys/tree/main/.github/actions/spell-check/allow

But do note that the spell checker only thinks about words that match its sense of words, so `nVidia` here will be ignored:
https://github.com/microsoft/PowerToys/blob/41f4d971dde3a5e5e818889bb0ad8e051c96086f/.github/actions/spell-check/allow/names.txt#L5

jsoref commented:
Fwiw, I'd personally prefer `IPv4` / `IPv6` over `IP v4`/`IP v6`.

htcfreek commented:
> Fwiw, I'd personally prefer `IPv4` / `IPv6` over `IP v4`/`IP v6`.

What do you mean?

jsoref commented:
There's a space between the literal `IP` and the literal `v4`/`v6`:
![image](https://user-images.githubusercontent.com/2119212/159047461-84e86c74-01c7-496a-b200-b75b889d82d3.png)

https://en.wikipedia.org/wiki/IPv4
https://en.wikipedia.org/wiki/IPv6

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System.UnitTests/QueryTests.cs:
```diff
@@ -27,6 +28,9 @@ public void Setup()
         [DataRow("sleep", "Put computer to sleep")]
         [DataRow("hibernate", "Hibernate computer")]
         [DataRow("empty recycle", "Empty Recycle Bin")]
+        [DataRow("ip v4 addr", "IP v4 address of")]
+        [DataRow("ip v6 addr", "IP v6 address of")]
```
```suggestion
        [DataRow("ip v4 addr", "IPv4 address of")]
        [DataRow("ip v6 addr", "IPv6 address of")]
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/NetworkConnectionProperties.cs:
```diff
@@ -0,0 +1,305 @@
+        internal string Suffix { get; private set; }
+
+        /// <summary>
+        /// Gets the IP v4 address
```
```suggestion
        /// Gets the IPv4 address
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/NetworkConnectionProperties.cs:
```diff
@@ -0,0 +1,305 @@
+        internal string IPv4 { get; private set; }
+
+        /// <summary>
+        /// Gets the IP v4 subnet mask
```
```suggestion
        /// Gets the IPv4 subnet mask
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/NetworkConnectionProperties.cs:
```diff
@@ -0,0 +1,305 @@
+        internal string IPv4Mask { get; private set; }
+
+        /// <summary>
+        /// Gets the primarily used IP v6 address
```
```suggestion
        /// Gets the primarily used IPv6 address
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/NetworkConnectionProperties.cs:
```diff
@@ -0,0 +1,305 @@
+        internal string IPv6Primary { get; private set; }
+
+        /// <summary>
+        /// Gets the global IP v6 address
```
```suggestion
        /// Gets the global IPv6 address
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/NetworkConnectionProperties.cs:
```diff
@@ -0,0 +1,305 @@
+        internal string IPv6Global { get; private set; }
+
+        /// <summary>
+        /// Gets the temporary IP v6 address
```
```suggestion
        /// Gets the temporary IPv6 address
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/NetworkConnectionProperties.cs:
```diff
@@ -0,0 +1,305 @@
+        internal string IPv6Temporary { get; private set; }
+
+        /// <summary>
+        /// Gets the link local IP v6 address
```
```suggestion
        /// Gets the link local IPv6 address
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/NetworkConnectionProperties.cs:
```diff
@@ -0,0 +1,305 @@
+        internal string IPv6LinkLocal { get; private set; }
+
+        /// <summary>
+        /// Gets the site local IP v6 address
```
```suggestion
        /// Gets the site local IPv6 address
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Components/NetworkConnectionProperties.cs:
```diff
@@ -0,0 +1,305 @@
+        internal string IPv6SiteLocal { get; private set; }
+
+        /// <summary>
+        /// Gets the unique local IP v6 address
```
```suggestion
        /// Gets the unique local IPv6 address
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Properties/Resources.resx:
```diff
@@ -144,6 +182,42 @@
+    <comment>'v4' is the protocol version.</comment>
+  </data>
+  <data name="Microsoft_plugin_sys_Ip4SubnetMask" xml:space="preserve">
+    <value>IP v4 subnet mask</value>
```
```suggestion
    <value>IPv4 subnet mask</value>
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Properties/Resources.resx:
```diff
@@ -144,6 +182,42 @@
+    <comment>'v4' is the protocol version.</comment>
+  </data>
+  <data name="Microsoft_plugin_sys_ip4_description" xml:space="preserve">
+    <value>IP v4 address of {0}</value>
```
```suggestion
    <value>IPv4 address of {0}</value>
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Properties/Resources.resx:
```diff
@@ -144,6 +182,42 @@
+    <comment>'v4' is the protocol version. Don't translate the placeholder '{0}' as it is replaced in code.</comment>
+  </data>
+  <data name="Microsoft_plugin_sys_Ip6Address" xml:space="preserve">
+    <value>IP v6 Address</value>
```
```suggestion
    <value>IPv6 Address</value>
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Properties/Resources.resx:
```diff
@@ -144,6 +182,42 @@
+    <comment>'v6' is the protocol version.</comment>
+  </data>
+  <data name="Microsoft_plugin_sys_Ip6Link" xml:space="preserve">
+    <value>IP v6 link local address</value>
```
```suggestion
    <value>IPv6 link local address</value>
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Properties/Resources.resx:
```diff
@@ -144,6 +182,42 @@
+    <comment>'v6' is the protocol version.</comment>
+  </data>
+  <data name="Microsoft_plugin_sys_Ip6Site" xml:space="preserve">
+    <value>IP v6 site local address</value>
```
```suggestion
    <value>IPv6 site local address</value>
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Properties/Resources.resx:
```diff
@@ -144,6 +182,42 @@
+    <comment>'v6' is the protocol version.</comment>
+  </data>
+  <data name="Microsoft_plugin_sys_Ip6Temp" xml:space="preserve">
+    <value>IP v6 temporary address</value>
```
```suggestion
    <value>IPv6 temporary address</value>
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Properties/Resources.resx:
```diff
@@ -144,6 +182,42 @@
+    <comment>'v6' is the protocol version.</comment>
+  </data>
+  <data name="Microsoft_plugin_sys_Ip6Unique" xml:space="preserve">
+    <value>IP v6 unique local address</value>
```
```suggestion
    <value>IPv6 unique local address</value>
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System/Properties/Resources.resx:
```diff
@@ -144,6 +182,42 @@
+    <comment>'v6' is the protocol version.</comment>
+  </data>
+  <data name="Microsoft_plugin_sys_ip6_description" xml:space="preserve">
+    <value>IP v6 address of {0}</value>
```
```suggestion
    <value>IPv6 address of {0}</value>
```

jsoref reviewed (commented):
It'd be _something_ like this, 

plus adding this to `patterns.txt` (from https://github.com/check-spelling/check-spelling/wiki/Configuration-Examples:-patterns)

```
# version suffix <word>v#
[Vv]\d+(?:\b|(?=[a-zA-Z_]))
```

htcfreek commented:
@jaimecbernardo
Wen do we close v0.57.0 for new features?

htcfreek commented:
@jsoref
I have updates the naming and added the pattern.

@stefansjfw 
I have changed scoring to give the results a lower priority. The user can disable this behavior in settings now. The Description of this setting can be moved into a secondary text when #17108 is merged.

niels9001 commented:
![networkAdapter light](https://user-images.githubusercontent.com/9866362/159128241-ecf34709-d6db-44fa-99bf-76e9674c32ce.png)
![networkAdapter dark](https://user-images.githubusercontent.com/9866362/159128244-b04e0b22-af51-43b7-aaf3-95f2926801e3.png)

@htcfreek Here you go :)

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System.UnitTests/ImageTests.cs:
```diff
@@ -28,6 +28,9 @@ public void Setup()
         [DataRow("hibernate", "Images\\sleep.dark.png")]
         [DataRow("empty recycle", "Images\\recyclebin.dark.png")]
         [DataRow("uefi firmware settings", "Images\\firmwareSettings.dark.png")]
+        [DataRow("ip v4 addr", "Images\\networkAdapter.dark.png")]
+        [DataRow("ip v6 addr", "Images\\networkAdapter.dark.png")]
```
I'm not sure what these are (sorry, I haven't tried to read C# in a while)

htcfreek replied:
These are some tests. Left side is the user inputbthat will be simulated. And right side is the image path we expected for the result.

jsoref commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.System.UnitTests/QueryTests.cs:
```diff
@@ -27,6 +28,11 @@ public void Setup()
         [DataRow("empty recycle", "Empty Recycle Bin")]
+        [DataRow("ip", "IPv4 address of")]
+        [DataRow("address", "IPv4 address of")] // searching for address should show ipv4 first
+        [DataRow("ip v4", "IPv4 address of")]
+        [DataRow("ip v6", "IPv6 address of")]
```
Similarly

htcfreek replied:
Same here. These are tests too.

jsoref reviewed (commented):
@htcfreek: looks good,

fwiw, you no longer need `IPv` in `expect` (and can also apparently remove `settingsv` which probably also was matched by the pattern):
https://github.com/microsoft/PowerToys/runs/5612548869?check_suite_focus=true#step:4:70
(and you can remove the other two items as well, or wait for someone else to...)

htcfreek commented:
> @htcfreek: looks good,
> 
> fwiw, you no longer need `IPv` in `expect` (and can also apparently remove `settingsv` which probably also was matched by the pattern):
> https://github.com/microsoft/PowerToys/runs/5612548869?check_suite_focus=true#step:4:70
> (and you can remove the other two items as well, or wait for someone else to...)

I think the patterns are bad if they mark things as okay that don't have anything to do with `IPv4/IPv6`.

jsoref commented:
Nah, the parsing is right `settings`+`v#` and `settings` is a word in the dictionary.

My guess is that both David and Morton are in the base dictionary and thus don't need to be in expect. (I'd have to check.)

htcfreek commented:
> Nah, the parsing is right `settings`+`v#` and `settings` is a word in the dictionary.
> 
> My guess is that both David and Morton are in the base dictionary and thus don't need to be in expect. (I'd have to check.)

Than you have updated base dict.

jaimecbernardo commented:
> @jaimecbernardo Wen do we close v0.57.0 for new features?

No set date yet, but most likely it should be mid-end of this week.

htcfreek commented:
> > @jaimecbernardo Wen do we close v0.57.0 for new features?
> 
> No set date yet, but most likely it should be mid-end of this week.

Currently I am waiting for a new approval from @stefansjfw and then this can go in.

stefansjfw reviewed (approved):
Looks good. Although, I'm not completely sure about "Reduce score" settings. But I guess we can adapt in the future if needed - ip and mac results are still on top of other plugins which could make some users unhappy. But, as I said, let's see..

htcfreek commented:
> Looks good. Although, I'm not completely sure about "Reduce score" settings. But I guess we can adapt in the future if needed - ip and mac results are still on top of other plugins which could make some users unhappy. But, as I said, let's see..

What do you searched? I tried my best. But this is a fuzzyMatch scoring thing.

What is you worry about the setting?

stefansjfw commented:
> > Looks good. Although, I'm not completely sure about "Reduce score" settings. But I guess we can adapt in the future if needed - ip and mac results are still on top of other plugins which could make some users unhappy. But, as I said, let's see..
> 
> What do you searched? I tried my best. But this is a fuzzyMatch scoring thing.
> 
> What is you worry about the setting?

That we'll end up with more "Reduce score" settings, and then scoring won't make sense :)
It feels to me that if user wants to get Ip or mac, it should be strict clear search as the intention is clear.
Something like adding prefix - "network ip" or "net ip".

htcfreek commented:
> > > Looks good. Although, I'm not completely sure about "Reduce score" settings. But I guess we can adapt in the future if needed - ip and mac results are still on top of other plugins which could make some users unhappy. But, as I said, let's see..
> > 
> > What do you searched? I tried my best. But this is a fuzzyMatch scoring thing.
> > 
> > What is you worry about the setting?
> 
> That we'll end up with more "Reduce score" settings, and then scoring won't make sense :)
> It feels to me that if user wants to get Ip or mac, it should be strict clear search as the intention is clear.
> Something like adding prefix - "network ip" or "net ip".  

What about `Network: IPv...
` and `Network adapter: MAC...`? This might help.

stefansjfw commented:
I don't have strong feeling about this. It's more of a suggestion. @crutkas @jaimecbernardo WDYT?

htcfreek commented:
@jaimecbernardo 
It's your turn to decide if this can go in or if we should adjust again.

jaimecbernardo reviewed (approved):
Just gave it a try. It feels OK to me. Thanks for the contribution.

jaimecbernardo commented:
It's ok to return these results in a good position from just "ip" or "mac". I like that the setting to reduce is on by default as well. Let's see how people feel about these.
